tests/test_infer.py

```diff
@@ -1064,3 +1065,62 @@ def g(c):
         else:
             return f3
     return g(x < 42)(x)
+
+
+@infer(type=[(i64, i64)],
+       shape=[({'type': ai64, 'shape': (6, 13)}, (6, 13))])
+def test_identity(x):
+    return identity(x)
+
+
+@infer(type=[(B, B, B),
+             (i64, B, InferenceError),
+             (B, i64, InferenceError)])
+def test_bool_and(x, y):
+    return bool_and(x, y)
+
+
+@infer(type=[(B, B, B),
+             (i64, B, InferenceError),
+             (B, i64, InferenceError)])
+def test_bool_or(x, y):
+    return bool_or(x, y)
+
+
+@infer(
+    type=[
+        (B, i64, i64, i64),
+        (i64, i64, i64, InferenceError),
+        (B, i64, f64, InferenceError),
+        ({'value': True}, i64, f64, i64),
+        ({'value': False}, i64, f64, f64),
+    ],
+    value=[
+        (True, 1, 2, 1),
+        (False, 1, 2, 2),
+        (ANYTHING, 1, 2, ANYTHING),
+    ],
+    shape=[
+        ({'type': B},
+         {'type': ai64, 'shape': (6, 13)},
+         {'type': ai64, 'shape': (6, 13)},
+         (6, 13)),
+
+        ({'type': B},
+         {'type': ai64, 'shape': (6, 13)},
+         {'type': ai64, 'shape': (6, 14)},
+         InferenceError),
+
+        ({'type': B, 'value': True},
+         {'type': ai64, 'shape': (6, 13)},
+         {'type': ai64, 'shape': (6, 14)},
+         (6, 13)),
+
+        ({'type': B, 'value': False},
+         {'type': ai64, 'shape': (6, 13)},
+         {'type': ai64, 'shape': (6, 14)},
+         (6, 14)),
```

Comment:
I'm not sure if we want to allow for shape mismatch.  I think it might be a better idea if both sides are required to have the same shape.

Also, there are no tests where one of the branches has `ANYTHING` as a shape element.

Reply:
I suppose. I'd tend to use the same behavior as for type inference, though, whichever it is.

Reply:
Yes, I agree, we should keep the same behaviour as type inference.